### PR TITLE
Verify column name length in PostgreSQL and SQL Server

### LINKED ISSUE #12882: PostgreSQL connector truncates long identifiers silently
```sql
# schema name
trino> CREATE SCHEMA postgresql.a123456789012345678901234567890123456789012345678901234567890123;
trino> SHOW SCHEMAS IN postgresql;
                             Schema
-----------------------------------------------------------------
 a12345678901234567890123456789012345678901234567890123456789012

# table name
trino> CREATE TABLE postgresql.tpch.a123456789012345678901234567890123456789012345678901234567890123 (c1 int);
trino> SHOW TABLES IN postgresql.tpch;
                              Table
-----------------------------------------------------------------
 a12345678901234567890123456789012345678901234567890123456789012

# column name
trino> CREATE TABLE postgresql.tpch.test (a123456789012345678901234567890123456789012345678901234567890123 int);
trino> DESC postgresql.tpch.test;
                             Column                              |  Type   | Extra | Comment
-----------------------------------------------------------------+---------+-------+---------
 a12345678901234567890123456789012345678901234567890123456789012 | integer |       |
```

https://www.postgresql.org/docs/current/limits.html
>  identifier length | 63 bytes | can be increased by recompiling PostgreSQL

I think we should block such identifiers using:
* DatabaseMetaData#getMaxCatalogNameLength
* DatabaseMetaData#getMaxSchemaNameLength
* DatabaseMetaData#getMaxTableNameLength
* DatabaseMetaData#getMaxColumnNameLength

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Fixes #12882

## Documentation

(x) No documentation is needed.

## Release notes

(x) Release notes entries required with the following suggested text:

```markdown
# PostgreSQL
* Prevent using column names with a name which is longer
  than the max length supported by PostgreSQL. Previously, long names were
  truncated to the max length. ({issue}`13742`)
# SQL Server
* Prevent renaming a column to a name which is longer
  than the max length supported by SQL Server. Previously, long names were
  truncated to the max length. ({issue}`13742`)
```

### CODE REVIEW

**Comment by findepi** [comment]
@ebyhr can you please rebase? there are conflicts.

**Comment by ebyhr** [comment]
Rebased on upstream to resolved conflicts.

File: plugin/trino-mariadb/src/main/java/io/trino/plugin/mariadb/MariaDbClient.java [hunk 1]
```diff
@@ -150,6 +150,8 @@
     // An empty character means that the table doesn't have a comment in MariaDB
     private static final String NO_COMMENT = "";
 
+    private static final int PARSE_ERROR = 1064;
```
Comment (findepi) [inline plugin/trino-mariadb/src/main/java/io/trino/plugin/mariadb/MariaDbClient.java:153]: is it documented? can you link to docs?

File: plugin/trino-mariadb/src/main/java/io/trino/plugin/mariadb/MariaDbClient.java [hunk 2]
```diff
@@ -457,7 +459,7 @@ public void renameColumn(ConnectorSession session, JdbcTableHandle handle, JdbcC
             execute(connection, sql);
         }
         catch (TrinoException e) {
-            if (e.getCause() instanceof SQLSyntaxErrorException) {
+            if (e.getCause() instanceof SQLSyntaxErrorException syntaxError && syntaxError.getErrorCode() == PARSE_ERROR) {
```
Comment (findepi) [inline plugin/trino-mariadb/src/main/java/io/trino/plugin/mariadb/MariaDbClient.java:462]: > SQLSyntaxErrorException contains invalid column name.

add a code comment 

```
// Note: SQLSyntaxErrorException can be throw also when column name is invalid
```

BTW do we need this complicated logic?
why not "drop support" for the older MariaDB version and simplify this?

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java [hunk 3]
```diff
@@ -779,18 +779,23 @@ public void renameColumn(ConnectorSession session, JdbcTableHandle handle, JdbcC
...
         }
     }
 
+    protected String renameColumnSql(JdbcTableHandle handle, JdbcColumnHandle jdbcColumn, String newRemoteColumnName)
+    {
+        return format(
+                "ALTER TABLE %s RENAME COLUMN %s TO %s",
+                quoted(handle.asPlainTable().getRemoteTableName()),
+                jdbcColumn.getColumnName(),
+                newRemoteColumnName);
```
Comment (findepi) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java:799]: Why aren't these quoted?

(i know they weren't quoted before the refactor)

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java [hunk 4]
```diff
@@ -1030,6 +1030,17 @@ protected void verifyTableName(DatabaseMetaData databaseMetadata, String tableNa
...
     }
 
+    @Override
+    protected void verifyColumnName(DatabaseMetaData databaseMetadata, String columnName)
+            throws SQLException
+    {
+        // PostgreSQL truncates table name to 63 chars silently
+        // PostgreSQL driver caches the max column name length in a DatabaseMetaData object. The cost to call this method per column is low.
+        if (columnName.length() > databaseMetadata.getMaxColumnNameLength()) {
+            throw new TrinoException(NOT_SUPPORTED, format("Column name must be shorter than or equal to '%s' characters but got '%s'", databaseMetadata.getMaxColumnNameLength(), columnName.length()));
```
Comment (findepi) [inline plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:1040]: Quote the contents of the `columnName` in the message besides their length.

It's important for CREATE TABLE statement, where there can be multiple potentially problematic columns in play

File: plugin/trino-sqlserver/src/main/java/io/trino/plugin/sqlserver/SqlServerClient.java [hunk 5]
```diff
@@ -288,6 +288,17 @@ protected void verifyTableName(DatabaseMetaData databaseMetadata, String tableNa
...
     }
 
+    @Override
+    protected void verifyColumnName(DatabaseMetaData databaseMetadata, String columnName)
+            throws SQLException
+    {
+        // SQL Server truncates table name to the max length silently when renaming a column
+        // SQL Server driver doesn't communicate with a server in getMaxColumnNameLength. The cost to call this method per column is low.
+        if (columnName.length() > databaseMetadata.getMaxColumnNameLength()) {
+            throw new TrinoException(NOT_SUPPORTED, format("Column name must be shorter than or equal to '%s' characters but got '%s'", databaseMetadata.getMaxColumnNameLength(), columnName.length()));
```
Comment (findepi) [inline plugin/trino-sqlserver/src/main/java/io/trino/plugin/sqlserver/SqlServerClient.java:298]: Quote the contents of the `columnName` in the message besides their length

**Review by findepi — APPROVED** [review]

File: plugin/trino-mariadb/src/main/java/io/trino/plugin/mariadb/MariaDbClient.java (hunk 2 above)
Comment (ebyhr) [reply plugin/trino-mariadb/src/main/java/io/trino/plugin/mariadb/MariaDbClient.java:464]: I think throwing an MariaDB exception without conversion is also fine. Let me handle in a follow-up PR.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java (hunk 3 above)
Comment (ebyhr) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java:799]: Thanks for catching that. Let me fix in a follow-up PR.

**Comment by ebyhr** [comment]
CI hit #13825

**Comment by hashhar** [comment]
Today while in looking into something related I saw a very disturbing behaviour from postgres. Directly in Postgres:

```
tpch=# select * from (select name as zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz, regionkey as zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz from tpch.region) t;
NOTICE:  identifier "zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz" will be truncated to "zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz"
NOTICE:  identifier "zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz" will be truncated to "zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz"
 zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz | zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz
-----------------------------------------------------------------+-----------------------------------------------------------------
 AFRICA                                                          |                                                               0
 AMERICA                                                         |                                                               1
 ASIA                                                            |                                                               2
 EUROPE                                                          |                                                               3
 MIDDLE EAST                                                     |                                                               4
(5 rows)

tpch=# select zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz from (select name as zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz, regionkey as zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz from tpch.region) t;
NOTICE:  identifier "zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz" will be truncated to "zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz"
NOTICE:  identifier "zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz" will be truncated to "zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz"
ERROR:  column reference "zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz" is ambiguous
LINE 1: select zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz...
               ^
tpch=# select zzzzzzzzzzzzzzzzzzzzzzzzzzzzz from (select name as zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz, regionkey as zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz from tpch.region) t;
NOTICE:  identifier "zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz" will be truncated to "zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz"
NOTICE:  identifier "zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz" will be truncated to "zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz"
ERROR:  column "zzzzzzzzzzzzzzzzzzzzzzzzzzzzz" does not exist
LINE 1: select zzzzzzzzzzzzzzzzzzzzzzzzzzzzz from (select name as zz...
```

This means that Postgres truncates ALL occurences of long identifiers and the truncated values are treated equivalently.
i.e. If postgres only allowed 2 char identifiers then `zzz` == `zzzz` == `zzzzzzz` and any other combination.

However when selecting from a relation which produces `zzzz` and `zzz` both get truncated to `zz` and the SELECT complains about ambiguity.

So it seems we're still "safe" in the sense that if for example join pushdown caused truncated identifiers then the worst thing that can happen would be query failure because postgres would complain of ambiguity.

**Pull request merged**